volume/store/store.go

```diff
@@ -268,27 +270,45 @@ func (s *VolumeStore) list() ([]volume.Volume, []string, error) {
 
 // CreateWithRef creates a volume with the given name and driver and stores the ref
 // This ensures there's no race between creating a volume and then storing a reference.
-func (s *VolumeStore) CreateWithRef(name, driverName, ref string, opts, labels map[string]string) (volume.Volume, error) {
+// In case a volume has already been created with the same name and driver, the already
+// created volumes will be returned
+func (s *VolumeStore) CreateWithRef(name, driverName, ref string, opts, labels map[string]string, found ...func(v volume.Volume) error) (volume.Volume, error) {
 	name = normaliseVolumeName(name)
 	s.locks.Lock(name)
 	defer s.locks.Unlock(name)
 
-	v, err := s.create(name, driverName, opts, labels)
+	v, err := s.create(name, driverName, opts, labels, found...)
 	if err != nil {
 		if _, ok := err.(*OpErr); ok {
-			return nil, err
+			return v, err
 		}
-		return nil, &OpErr{Err: err, Name: name, Op: "create"}
+		return v, &OpErr{Err: err, Name: name, Op: "create"}
 	}
 
 	s.setNamed(v, ref)
 	return v, nil
 }
 
 // Create creates a volume with the given name and driver.
-// This is just like CreateWithRef() except we don't store a reference while holding the lock.
+// In case a volume has already been created with the same name and driver, the already
+// created volumes will be returned together with the errAlreadyExists error.
 func (s *VolumeStore) Create(name, driverName string, opts, labels map[string]string) (volume.Volume, error) {
-	return s.CreateWithRef(name, driverName, "", opts, labels)
+	return s.CreateWithRef(name, driverName, "", opts, labels, func(vol volume.Volume) error {
```

Comment:
This is the only place we're using this "found" and it seems legit to put this in the volume conflict lookup check.